packages/labs/forms/src/form-associated.ts

```diff
@@ -0,0 +1,227 @@
+    }
+
+    formDisabledCallback(disabled: boolean) {
+      if (disabled) {
+        disabledElements.add(this);
+      } else {
+        disabledElements.delete(this);
+      }
+      this.#internals.ariaDisabled = String(disabled);
+      this.requestUpdate();
+    }
+
+    formResetCallback() {
+      const metadata = (this.constructor as FormAssociatedConstructor)[
+        Symbol.metadata
+      ];
+
+      // Restore the initial value
+      const valueAccess = valueAccessors.get(metadata);
+      const initialValue = initialValues.get(this);
+      valueAccess?.set(this, initialValue);
+
+      // Restore the initial state
+      const stateAccess = stateAccessors.get(metadata);
+      const initialState = initialStates.get(this);
+      stateAccess?.set(this, initialState);
+    }
+
+    formStateRestoreCallback(
+      state: string | File | FormData | null,
+      mode: 'restore' | 'autocomplete'
+    ) {
+      const metadata = (this.constructor as FormAssociatedConstructor)[
+        Symbol.metadata
+      ];
+      if (mode === 'restore') {
+        // When restoring a value we should get the state previously passed to
+        // internals.setFormValue(). We can pass that to the state accessor,
+        // which should update the value accessor.
+        // If there is no state accessor, we should update the value accessor
+        const stateAccess = stateAccessors.get(metadata);
+        if (stateAccess !== undefined) {
+          stateAccess.set(this, state);
+        } else {
+          const valueAccess = valueAccessors.get(metadata);
+          valueAccess?.set(this, state);
+        }
+      } else if (mode === 'autocomplete') {
+        // When autocompleting a value, we don't get a previous state object, so
+        // we should update the value accessor directly.
+        const valueAccess = valueAccessors.get(metadata);
+        valueAccess?.set(this, state);
+      }
+    }
+  }
+  return C as Constructor<Interface<FormAssociated>> & T;
+};
+
+type Access = ClassAccessorDecoratorContext<FormAssociated, unknown>['access'];
+
+const valueAccessors = new WeakMap<DecoratorMetadataObject, Access>();
+
+/*
+ * A store of initial values for FormAssociated elements so that they can be
+ * restored when the form is reset.
+ */
+const initialValues = new WeakMap<FormAssociated, unknown>();
+
+/**
+ * A class accessor decorator that marks a field as the form value for the
+ * FormAssociated mixin.
+ */
+export const formValue =
+  () =>
+  <V extends string | File | FormData | null>(
+    target: ClassAccessorDecoratorTarget<FormAssociated, V>,
+    context: ClassAccessorDecoratorContext<FormAssociated, V>
+  ): ClassAccessorDecoratorResult<FormAssociated, V> => {
+    valueAccessors.set(context.metadata, context.access);
+
+    return {
+      get: target.get,
+      set(value: V) {
+        target.set.call(this, value);
+        getInternals(this)!.setFormValue(value);
+      },
+      init(value: V) {
```

Comment:
So this requires using standard decorators. Think that's ok?

Reply:
Just sketching things out for now and making sure they compile. I think we'll need to support experimental decorators though, and figure out how to handle the default value. Maybe we try out the pattern we're discussion for `@property()` here first.

Reply:
If we need user-configurable default values, then we'll have to have a separate decorator for defaults and won't need to rely on the initializer anymore.